keps/sig-node/2400-node-swap/README.md

```diff
@@ -0,0 +1,707 @@
+
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+-->
+
+It is possible for this feature to affect performance of some worker node-level SLIs/SLOs. We will need to monitor for differences, particularly during beta testing, when evaluating this feature for beta and graduation.
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+<!--
+Things to keep in mind include: additional in-memory state, additional
+non-trivial computations, excessive access to disks (including increased log
+volume), significant amount of data sent and/or received over network, etc.
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+Yes. It will permit the utilization of swap memory (i.e. disk) on nodes. This is expected, as this enhancement is enabling cluster administrators to access this resource.
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+- **2015-04-24:** Discussed in [#7294](https://github.com/kubernetes/kubernetes/issues/7294).
+- **2017-10-06:** Discussed in [#53533](https://github.com/kubernetes/kubernetes/issues/53533).
+- **2021-01-05:** Initial design discussion document for swap support and use cases.
+- **2021-04-05:** Alpha KEP drafted for initial node-level swap support and implementation (KEP-2400).
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+When swap is enabled, particularly for workloads, the kubelet’s resource accounting may become much less accurate. This may make cluster administration more difficult and less predictable.
```

Comment:
> When swap is enabled, particularly for workloads

What exactly does this mean? The `particularly for workloads` seems a bit confusing? Does this mean that it's less of a concern on nodes where... we **aren't** running workloads? 

> This may make cluster administration more difficult and less predictable.

I am assuming this is because of how SWAP works with Linux and the dynamic nature of exchanging runtime in memory resources with disk arbitrarily. Do we have any tests or examples of when this has caused any dramatic changes?

Also given the upcoming (and potentially unreliable) [changes in Linux 5.12](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=caf6912f3f4af7232340d500a4a2008f81b93f14) do we have any warnings/alarms for if someone is using SWAP?

It would be a shame to be running a container runtime on a host system with SWAP enabled and be liable for both kernel level deviations as well as container runtime level deviations.

Reply:
> What exactly does this mean? The `particularly for workloads` seems a bit confusing? Does this mean that it's less of a concern on nodes where... we **aren't** running workloads?

By default, even if you let the kubelet run with swap on, this proposal disables any swap being allocated to Kubernetes workloads via the CRI. But you can turn that on if you like with the `KubeConfig.MemorySwapLimit` config.

> I am assuming this is because of how SWAP works with Linux and the dynamic nature of exchanging runtime in memory resources with disk arbitrarily. Do we have any tests or examples of when this has caused any dramatic changes?

Nope, we need to gather the data and metrics for this!

> Also given the upcoming (and potentially unreliable) [changes in Linux 5.12](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=caf6912f3f4af7232340d500a4a2008f81b93f14) do we have any warnings/alarms for if someone is using SWAP?
> 
> It would be a shame to be running a container runtime on a host system with SWAP enabled and be liable for both kernel level deviations as well as container runtime level deviations.

We're not changing any of the defaults re: swap, so cluster operators must 1) provision it 2) (for alpha) turn on the feature flag 3) set fail-on-swap to false 4) if desired, tweak the KubeletConfig to allow kube workloads to use swap. There's no way to accidentally use it, and my goal is to give lots of flexibility in how it can be used.

Reply:
Nice! Thanks for the lovely responses @ehashman! This PR is great - I am so stoked to check it out. 

 > Nope, we need to gather the data and metrics for this!

Please do look me up if there is ever any work needed for getting metrics out of the kernel in an easy to use manner.